### ISSUE TITLE
Modeling of language independent text

### ISSUE THREAD

**Issue by DavidGrant-NIWC** [iho-ohi/S-101-Documentation-and-FC#60]
@alvarosanuy asked me to open this issue based on discussions in the S-101 portrayal sub-working group. The relevant comment is 
[PSWG Issue 104 comment](https://github.com/S-101-Portrayal-subWG/Working-Documents/issues/104#issuecomment-1452956621) (also copied below), but the entire thread should be read for context [PSWG Issue 104](https://github.com/S-101-Portrayal-subWG/Working-Documents/issues/104)

I believe the ask is that the DCEG re-consider the modeling of language independent text to better match the S-52 requirements and to simplify mariner selection of national language text. Some proposed modeling changes will be presented in a follow-on comment.

# Portrayal sub-working group issue comment
> He feels (and we all agreed), more discussion is required at the S100WG level and detailed performance instructions documented in the relevant instruments. Note that the use of language is bigger than featureName and also impacts the feature Information (S-57 ; TXTDSC, NTXTDS, INFORM and NNINFOM).

This is an S-101 issue, not an S-100 issue. I believe the root issue with national language text in S-101 is that for some reason the S-57 model was changed for S-101.

### S-57
In S-57 my understanding is that the (optional) `National language` independent mariner selector is used to toggle between OBJNAM and NOBJNM (text group 31), Specific implementation details are left up to each OEM. While a given chart could theoretically have text in more than two languages, a given feature can only have English and at most one national language equivalent. `National language` also must be enabled to show NTXTDS information in the pick report (implementation details are again left up to the OEM).

### S-64
![image](https://user-images.githubusercontent.com/62906211/222626315-956bdd37-f09e-47a0-a0d5-7733729c08c0.png)

### S-101
My understanding is that S-101 data should be encoded as follows to model the indicated purpose (edit Mar 3 to add OBJNAM/NOBJNM):
|Purpose|S-57|S-101|
|:---|:---|:---|
|International object/feature name|OBJNAM|`featureName.name` and `featureName.language`=(`unknown` or `eng`)|
|National language object name|NOBJNM|`featureName.name` and `featureName.language` is not (`unknown` or `eng`)|
|International text|INFORM|`information.text` and `information.language`=(`unknown` or `eng`)|
|National language text|NINFOM|`information.text` and `information.language` is not (`unknown` or `eng`)|
|International text in support file|TXTDSC|`information.fileReference` and `information.language`=(`unknown` or `eng`)|
|National language text in support file|NTXTDS|`information.fileReference` and `information.language` is not (`unknown` or `eng`) [this should probably be better described in the DCEG...]|
|Picture in support file|PICREP|`pictorialRepresentation`|

The issue is that S-101 allows features to encode an unlimited number of national language strings (note the multiplicity on the `information` binding):
![image](https://user-images.githubusercontent.com/62906211/222622459-8df4be8e-020b-439e-ac45-5fc752b1dc2f.png)

It also allows multiple national language strings by supporting multiple methods of adding `information` to a feature:
* via attribute binding to complex attribute `information`
* via information association `AdditionalInformation`

For example, `Berth` provides an `information` attribute and permits `AdditionalInformation` (`ContactDetails`, `NauticalInformation`, `NonStandardWorkingDay`, `ServiceHours`)

I understood this support for multiple different national language representations of a single international string to be intentional, but maybe it wasn't. If the model was changed to match S-52/S-57 (a single international text string is only allowed to have one national language equivalent) then we could change the `National language` context parameter from a string (`eng`) to a boolean and make things easier on ourselves and the mariner.

If this change were to be made, the mariner wouldn't have to know to type `spa` to see the national language text; they would just enable the `National language` boolean selector. This would match the S-52 functionality as tested in the S-64 test above.

**Comment by DavidGrant-NIWC** [comment]
# Current Model
## Feature name / OBJNAM
|Attribute|S-57 Acronym|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`featureName`|| |C|typically 0,∞|
|&emsp;`displayName`||`true` or `false`|(S) BO|0,1|
|&emsp;`language`|<sup>*</sup>|ISO 639-2/T|(S) TE|0,1|
|&emsp;`name`|`OBJNAM`, `NOBJNM`<sup>*</sup>||(S) TE|1,1|

\* `language`=(`unknown` or `eng`) indicates that `name` models `OBJNAM`, otherwise `name` models `NOBJNM`
## Information / INFORM
|Attribute|S-57 Acronym|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`information`|| |C|typically 0,∞|
|&emsp;`fileLocator`|||(S) TE|0,1|
|&emsp;`fileReference`|`TXTDSC`, `NTXTDS`<sup>*</sup>||(S) TE|0,1<sup>**</sup>|
|&emsp;`headline`|||(S) TE|0,1|
|&emsp;`language`|<sup>*</sup>|ISO 639-2|(S) TE|0,1|
|&emsp;`text`|`INFORM`, `NINFOM`<sup>*</sup>||(S) TE|0,1<sup>*</sup>|

\* `language`=(`unknown` or `eng`) indicates modeling of `TXTDSC` or `INFORM`, otherwise `NTXTDS` or `NINFOM` is being modeled.
\*\* For each instance of `information`, at least one of the sub-attributes `fileReference` or `text` must be populated.

# Notional Model
This model allows long and short names, alternate text, and other features of the current S-101 model but restricts the ability to encode multiple national language values for a given international string value.

Note: We could add `language` attributes to describe the national language, but they would be strictly informational. There can only be a single (optional) translation of each international string.

## Language independent headline
|Attribute|Description|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`iso639Headline`|language independent headline| |C||
|&emsp;`headline`|international headline||(S) TE|1,1|
|&emsp;`nationalHeadline`|||(S) TE|0,1|

## Language independent text
|Attribute|Description|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`iso639Text`|language independent text| |C||
|&emsp;`iso639Headline`|||C|0,1<sup>*</sup>|
|&emsp;`text`|international text||(S) TE|0,1<sup>*</sup>|
|&emsp;`nationalText`|national text||(S) TE|0,1|

\* For each instance of `iso639Text`, at least one of the sub-attributes `iso639Headline` or `text` must be populated.

## Text reference
|Attribute|Description|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`textReference`|ref to text in a file| |C||
|&emsp;`fileReference`|file URI||(S) TE|1,1|
|&emsp;`fileLocator`|||(S) TE|0,1|
|&emsp;`iso639Headline`|||C|0,1|

## Feature name / OBJNAM
|Attribute|S-57 Acronym|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`featureName`|| |C|typically 0,∞|
|&emsp;`displayName`||`true` (default value) or `false`|(S) BO|0,1|
|&emsp;`iso639Text`|`OBJNAM`, `NOBJNM`||C|1,1|

## Information / INFORM
|Attribute|S-57 Acronym|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`information`|| |C|typically 0,∞|
|&emsp;`textReference`|`TXTDSC`, `NTXTDS`||C|0,1<sup>*</sup>|
|&emsp;`iso639Text`|`INFORM`, `NINFOM`||C|0,1<sup>*</sup>|

\* For each instance of `information`, at least one of the sub-attributes `textReference` or `iso639Text` must be populated.

## Online resource
|Attribute|S-57 Acronym|Encoding Value|Type|Multiplicity|
|:---|:---|:---|:---|:---:|
|`onlineResource`|| |C|typically 0,∞|
|&emsp;`linkage`|||(S) TE|1,1|
|&emsp;`iso639Text`|||C|0,1|

Other features could be made language independent in a similar manner.

**Comment by rmalyankar** [comment]
Changing attribute definitions, like re-defining headline as _international headline_  or re-modeling things like _onlineResource_  and _information_ will promote this from an "S-101 issue" to a registry-wide issue affecting all Product Specifications, and I would have to take this kind of revision for discussion in NIPWG, especially when the _headline_ attribute (for one) and _onlineResource_ originated with NIPWG and are widely used in NPUBs Product Specifications and any re-definition, re-purposing, or re-modeling will affect them all.

If the intent is to support only one language other than English, why not retain the current modeling of _featureName_ and _information_ and just restrict the multiplicity **in S-101** to 0..2 or 1..2?

**Comment by DavidGrant-NIWC** [comment]
> Changing attribute definitions, like re-defining headline as international headline or re-modeling things like onlineResource and information will promote this from an "S-101 issue" to a registry-wide issue affecting all Product Specifications, and I would have to take this kind of revision for discussion in NIPWG, especially when the headline attribute (for one) and onlineResource originated with NIPWG and are widely used in NPUBs Product Specifications and any re-definition, re-purposing, or re-modeling will affect them all.

Point taken but note that `onlineResource` does not currently support language independence. While we could modify the notional model to eliminate changes to `headline`, I don't see a way to address the issue without remodeling `information` (but perhaps someone has other ideas).

> If the intent is to support only one language other than English, why not retain the current modeling of `featureName` and `information` and just restrict the multiplicity in S-101 to 0..2 or 1..2?

That could work, but note:
* In the thread I linked @TomRichardson6 notes: _I believe the modelling was changed to be more flexible to allow both the full and shortened names to be used to improve the display for users and to allow the language to be specifically identified._
* Restricting to just two entries precludes referencing multiple sections of a support file via `fileLocator`, or referencing multiple support files.
* The DCEG description of `information` contains: _This complex attribute should be used, for example, to hold the information that is shown on paper charts by cautionary and explanatory notes_
  * I assume we need the ability to reference more than one cautionary / explanatory note from a given feature.

**Comment by DavidGrant-NIWC** [comment]
# Recommended Changes - 2nd try

This assumes that:
* We prefer to minimize changes to the existing model.
  * We are ok with providing special cases to handle each emergent portrayal requirement.
  * Other products will provide their own (possibly conflicting) handling for their own portrayal requirements.
* We only need to address language issues which impact the S-101 portrayal. 
  * Pick report / legend / etc. will be addressed separately, either in the PS or S-100.
    * At minimum, the PS or S-100 should provide guidance regarding language independence and use of the `National language` independent mariner selector.
  * The S-101 portrayal will not need to portray `INFORM` for any S-101 feature which didn't exist in S-57.
    * `sectorInformation`, `shapeInformation`, `information`, etc. do not need to be portrayed.
  * `onlineResource` and any other attributes which don't support language independence will be addressed separately (or not).

## Feature name
Add restrictions to ensure only one instance of `OBJNAM` and `NOBJNM`
|Attribute|S-57 Acronym|Encoding Value|Default|Type|Multiplicity|
|:---|:---|:---|:---:|:---|:---:|
|`featureName`|| ||C|typically 0,∞|
|&emsp;`displayName`<sup>1,2</sup>||`true` or `false`|`true`|(S) BO|0,1|
|&emsp;`language`||ISO 639-2/T|`eng`|(S) TE|0,1|
|&emsp;`name`|`OBJNAM`<sup>3</sup>, `NOBJNM`<sup>4</sup>|||(S) TE|1,1|

<sup>1</sup> `displayName`: must be set to `false` for all international names which do not model `OBJNAM` (there can only be one `OBJNAM`)
<sup>2</sup> `displayName`: must be set to `false` for all national names which do not model `NOBJNM` (there can only be one `NOBJNM`)
<sup>3</sup> `OBJNAM`: value of `name` when `displayName` is `null` or `true`, and `language` is `null` or `eng`
<sup>4</sup> `NOBJNM`: value of `name` when `displayName` is `null` or `true`, and `language` is not `null` or `eng`

## Information
No change, just clarifying S-57 equivalents. `INFORM`, `NINFOM`, `TXTDSC`, and `NTXTDS` are never shown in the portrayal, they are only shown in the pick report.
|Attribute|S-57 Acronym|Encoding Value|Default|Type|Multiplicity|
|:---|:---|:---|:---:|:---|:---:|
|`information`<sup>1</sup>|| ||C|typically 0,∞|
|&emsp;`fileLocator`||||(S) TE|0,1|
|&emsp;`fileReference`|`TXTDSC`<sup>2</sup>, `NTXTDS`<sup>3</sup>|||(S) TE|0,1|
|&emsp;`headline`||||(S) TE|0,1|
|&emsp;`language`||ISO 639-2|`eng`|(S) TE|0,1|
|&emsp;`text`|`INFORM`<sup>2</sup>, `NINFOM`<sup>3</sup>|||(S) TE|0,1|

<sup>1</sup> For each instance of `information`, at least one of the sub-attributes `fileReference` or `text` must be populated.
<sup>2</sup> When `language` is `null` or `eng`
<sup>3</sup> When `language` is not `null` or `eng`

## Traffic separation scheme lane part
Added `warningInformation` to support portrayal requirements (similar to `sectorInformation`, `shapeInformation`, etc.).
|Attribute|S-57 Acronym|Encoding Value|Default|Type|Multiplicity|
|:---|:---|:---|:---:|:---|:---:|
|`fixed date range`<sup>1</sup>||See clause 2.4.8||C|0,1|
|&emsp;`date end`|`DATEND`|||(S) TD|0,1<sup>1</sup>|
|&emsp;`date start`|`DATSTA`|||(S) TD|0,1<sup>1</sup>|
|`status`|`STATUS`|1,3,9,28||EN|0,∞|
|`scale minimum`|`SCAMIN`|See clause 2.5.9||IN|0,1|
|`information`<sup>2</sup>|| ||C|0,∞|
|&emsp;`fileLocator`||||(S) TE|0,1|
|&emsp;`fileReference`|`TXTDSC`<sup>3</sup>, `NTXTDS`<sup>4</sup>|||(S) TE|0,1|
|&emsp;`headline`||||(S) TE|0,1|
|&emsp;`language`||ISO 639-2|`eng`|(S) TE|0,1|
|&emsp;`text`|`INFORM`<sup>3</sup>, `NINFOM`<sup>4</sup>|||(S) TE|0,1|
|`warningInformation`<sup>5</sup>|| ||C|0,2|
|&emsp;`language`||ISO 639-2/T|`eng`|(S) TE|0,1|
|&emsp;`text`|`INFORM`<sup>3</sup>, `NINFOM`<sup>4</sup>|||(S) TE|1,1|

<sup>1</sup> For each instance of fixed date range, at least one of the sub-attributes date end or date start must be 
populated.
<sup>2</sup> For each instance of `information`, at least one of the sub-attributes `fileReference` or `text` must be populated.
<sup>3</sup> When `language` is `null` or `eng`
<sup>4</sup> When `language` is not `null` or `eng`
<sup>5</sup> Only allowed when the `TrafficSeparationSchemeLanePart` is a junction

# Supporting Notes
Further investigation revealed that S-52 only portrays `INFORM`/`NINFOM` for `DistanceMark` (`DISMAR`) and `TrafficSeparationSchemeLanePart` (`TSSLPT`).
* `DistanceMark` / `DISMAR`
  * The S-52 portrayal reads the `INFORM` attribute of `DISMAR` to get the distance value.
  * The S-101 portrayal reads the `measuredDistanceValue` attribute of `DistanceMark` to get the distance.
    * S-101 doesn't need to portray `INFORM` or `NINFOM` for `DistanceMark`
    * `measuredDistanceValue` doesn't need to be translated (it's a numerical value)
    * `distanceUnitOfMeasurement` can only be translated by a language pack
* `TrafficSeparationSchemeLanePart` / `TSSLPT`
  * S-57 `TSSLPT` uses `INFORM` or `TXTDSC` to encode warning text.
    *  The warning text is only shown when:
        *  the warning is encoded in `INFORM` (not in `TXTDSC`)
        * `ORIENT` is not present (the `TSSLPT` is a junction)
        ![image](https://user-images.githubusercontent.com/62906211/223209767-c5dfe12b-67d5-4539-afa2-0b3961ec3395.png)
        ![image](https://user-images.githubusercontent.com/62906211/223210254-bc6b12e6-2f39-47e4-9de9-90ccf03645b8.png)
  * S-101 `TrafficSeparationSchemeLanePart` uses `information` to encode warning text.
    ![image](https://user-images.githubusercontent.com/62906211/223211900-d3dd1d17-7114-4bb9-a759-d93cfdfe3b84.png)
    * It isn't possible for the portrayal to determine which `information` attribute provides the `INFORM` value intended to be portrayed and which provides the `NINFOM` value intended to be portrayed.
    * It's also not clear if those values could instead be provided through a `NauticalInformation` association

In order to support implementation of the S-101 portrayal without changing the modeling of `featureName` or `information` we could:
* Add restrictions to `featureName` to ensure `OBJNAM` and `NOBJNM` are singletons.
* Update the modeling for `TrafficSeparationSchemeLanePart` (`TSSLPT`) to support language independent encoding of warning text.

**Comment by rmalyankar** [comment]
Adding an optional _language_ attribute to _onlineResource_ would not disrupt other product specifications. OTOH multilingual web sites frequently have their own language selectors on site pages, so either the _language_ attribute might not be used as much as anticipated or would have to be 0..* to accommodate multiple languages on the same site.

The multiplicity of the _onlineResource_ complex attribute itself can be 0..*. Each instance of _onlineResource_ should indicate only a single resource. A different URL or an alternate site is a different resource and should be encoded in its own instance of _onlineResource_.

Is there a requirement to allow encoding of multiple feature names in the same language, i.e., with the same combination of_language_ and  _displayName_ values? If not, the constraints can probably be simplified. Maybe even define a new _screenName_ attribute and deprecate _displayName_ - that would not disrupt other product specifications and it would be easier to convince.

Given that product specifications are updated on different cycles, often multi-year cycles, accommodating disparate modeling will be unavoidable.

**Comment by DavidGrant-NIWC** [comment]
> [... `onlineResource` discussion ...]
* A separate issue should be opened for discussion and any proposed changes to `onlineResource`.

> Is there a requirement to allow encoding of multiple feature names in the same language, i.e., with the same combination of _language_ and _displayName_ values? [...]
* As I noted previously, that is my understanding based on comments from @TomRichardson6:
  * _In the thread I linked @TomRichardson6 notes: I believe the modelling was changed to be more flexible to allow both the full and shortened names to be used to improve the display for users and to allow the language to be specifically identified._

I need some resolution of this issue in order to implement national language support in the portrayal.

**Comment by rmalyankar** [comment]
> > [... `onlineResource` discussion ...]
> 
> * A separate issue should be opened for discussion and any proposed changes to `onlineResource`.

No objection to opening a separate issue. However, _onlineResource_ is mentioned upthread.

> 
> I need some resolution of this issue in order to implement national language support in the portrayal.

Recommend resolving this for implemention purposes by defining suitable constraints on the multiplicity and content of _featureName_, _information_, and their sub-attributes. This appears to be the way the discussion in the linked issue https://github.com/S-101-Portrayal-subWG/Working-Documents/issues/104  has gone too.

1. Require that each _language_+_displayName_ combination occur only once in any _featureName_'s that are bound to the same feature or info type, counting absent _language_ or _displayName_ as the defaults (cf. DCEG 27.87, modified to account for languages).
2. Require that if there is an "English" instance of _featureName_+_displayName_ combination in a feature, there should be a corresponding instance of _featureName_ for every other language used in the dataset.
3. Consider limiting the number of additional languages in a dataset to a small number (1 or 2 additional languages?), for convenience.
4. If a dataset contains attributes encoded in languages other than English, all the languages used must be encoded in the corresponding discovery metadata block in the exchange catalogue (_defaultLocale_ and _otherLocale_). The system can obtain the list of languages and codes from the dataset discovery block(s) and construct a list of available languages to show to the user. Language codes and names are in the S-100 resources file https://schemas.s100dev.net/schemas/S100/5.0.0/resources/Codelists/cat/codelists.xml, codelist _S100_MD_LanguageCode_, and that file or a transformation of it into an implementation-specific equivalent should be included on the system. It can be transferred as a support file in exchange set(s) if desired.
5. Languages in datasets should be limited to those in _S100_MD_LanguageCode_. If a language needs to be added to that list let me know ASAP.

**Comment by MikkoHovi-FI** [comment]
I finally tried to get my head around this issue, and noticed that modelling of languages, and names and texts has become a bit different than I remembered. This might be the reason why I quite haven't understood what it is that we are trying to solve here. 

It might be me remembering it incorrectly, but **my understanding** from before regarding languages, names and texts is that they were intended to work as follows:

1. Concept of 'national language' is dropped. The issue with S-57 is that it assumes that there is _one_ national language. This is not always true. Finland has two national languages, neither of them English, and in one specific region *five* non-English official languages - all treated equally in that region. (And, yes, we do publish charts from there.) To fix this the names were changed to complex attributes and the language attribute was introduced in the S-101 modelling.
2. S-100 ECDIS may offer the users to define their own preferred language(s) in its settings. This functionality would be similar as that found in many modern operating systems and software. This functionality lets the user to specify the language or languages that they want to use and in what order of preference.
3. ECDIS would portray only the feature name in the language which has the highest priority in the language preference list of the user. If none of the languages for the feature name matches with the preference list or no such preference exists, ECDIS will use the English content. Other languages than the one displayed would be available through pick report. For textual information and files the logic would be similar - the text or file with the highest priority would be shown or listed first , but the other languages would still be available through pick report etc.
5. As English would be considered either the name or text with language 'eng' or null, and there must only one of these.
6. Names are usually not really in English. It's one of local languages that is used in English instead. And this is where my recollection differs significantly from the current DCEG definitions: the 'display name' boolean was introduced to indicate which of the non-English names would be the "English name" in case it was needed. 'Display name' set to true would in this case be basically the same "use as the English name". This would then require constraints and checks to control the use of language = 'eng'/null and display name = true.

With this approach I could encode the following:

```
feature name
- display name = true
- language = fin
- name = Helsinki
feature name
- language = swe
- name = Helsingfors
```
This would portray as Helsinki unless there exists a user specified language preference list with Swedish in it and that Swedish also higher up than Finnish or English in it. In this case this would portray as Helsingfors.

This enables the user to set the language they want (provided that the data supports it) and not just a local "national language" which might actually be more than one language.

**Comment by DavidGrant-NIWC** [comment]
I think the primary issue could be resolved by allowing `featureName.displayName` = `true` for a single instance of each value of `language`. In other words, a single label in each language may be indicated as intended for display. Currently the DCEG only allows one instance of `featureName` with `displayName` = `true`, so only the English language value can be portrayed.

There are other changes which could be made to help reduce clutter in the pick report, or to provide consistency in national language support, but those changes aren't as important.

A secondary concern is how to allow the user to select a national language in a "friendly" way, but that is not a DCEG issue.

**Comment by JeffWootton** [comment]
Refer to discussions at DCEG Sub-Group 4 meeting (September 2023), Paper S-101PT11-06.1, related decisions and Action S-101PT11-08. Changes applied for S-101 Edition 1.2.0 at DCEG clauses 2.4.6, 2.5.8, 29.2 and throughout the document as required. Proposed re-modelling discussed with relevant S-101PT Members in the margins of the S-101PT11 meeting.

Close Issue.

**Comment by TomRichardson6** [comment]
A test dataset currently aligned with S-101 1.1.0 has been created to support testing of this issue it will be updated to edition 1.2.0 in due course. 

https://github.com/iho-ohi/S-101-Test-Datasets/tree/main/S-101_Test_DataSets/cells/101AA00DS0001/3
